lib/strict-non-empty-containers/src/Data/Map/NonEmpty/Strict/Internal.hs

```diff
@@ -0,0 +1,190 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- Provides a strict implementation of a non-empty map.
+--
+-- This implementation is based on the implementation of 'Data.Map.Strict'
+-- provided by the 'containers' package, but provides an extra guarantee
+-- that the map contains at least one entry at all times.
+--
+module Data.Map.NonEmpty.Strict.Internal
+    (
+    -- * Map type
+
+      -- Important:
+      --
+      -- The default data constructor for 'NonEmptyMap' is not exported, by
+      -- design, as the internal data structure has an invariant that must be
+      -- preserved across all operations.
+      --
+      -- Exporting the default constructor would make it possible for functions
+      -- outside this module to break the invariant, opening the door to subtle
+      -- regressions.
+      --
+      -- See the definition of 'NonEmptyMap' for more details of the invariant.
+      --
+      -- To construct a 'NonEmptyMap', use one of the provided constructors,
+      -- all of which are tested to check that they respect the invariant.
+      --
+      NonEmptyMap
+
+    -- * Construction
+    , fromList
+    , fromMap
+    , singleton
+
+    -- * Deconstruction
+    , toList
+    , toMap
+
+    -- * Insertion
+    , insert
+
+    -- * Deletion
+    , delete
+
+    -- * Lookup
+    , lookup
+
+    -- * Combination
+    , unionWith
+
+    -- * Internal functions
+    , invariantHolds
+
+    ) where
+
+import Prelude hiding
+    ( lookup )
+
+import Control.DeepSeq
+    ( NFData )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Map.Strict
+    ( Map )
+import GHC.Generics
+    ( Generic (..) )
+
+import qualified Data.Foldable as F
+import qualified Data.Map.Strict as Map
+
+-- | A non-empty map from keys of type 'k' to values of type 'v'.
+--
+data NonEmptyMap k v = NonEmptyMap
+    { least
+        :: !(k, v)
+      -- ^ Invariant: this key must be less than than all the keys in 'rest'.
+    , rest
+        :: !(Map.Map k v)
+    }
+    deriving (Eq, Foldable, Functor, Generic, Show, Traversable)
```

Comment:
This representation seems to make everything harder.
What about wrapping a Map in newtype, and preserve your invariants in the functions which operate on the NonEmptyMap.

Reply:
> What about wrapping a Map in newtype, and preserve your invariants in the functions which operate on the NonEmptyMap.

That's an interesting idea. Would you mind if I experiment with that in a future PR?

The only (minor) downside I can think of is that we might have some "impossible" cases within pattern matches. But it might be simpler overall.

The published non-empty (lazy) map types all use the "least" element technique, which I borrowed here. (For example: [nonempty-containers](https://hackage.haskell.org/package/nonempty-containers-0.3.4.1/docs/src/Data.Map.NonEmpty.Internal.html#NEMap).)